### PR TITLE
Initial set of guidelines for large scale training for Computer Vision

### PR DESCRIPTION
**Pull request by rdondera-microsoft**
# Description

Provide basic guidelines and sample code for large scale training of Computer Vision models. Future PRs will cover techniques such as DeepSpeed, ORT etc.

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/Azure/azureml-examples/blob/main/CONTRIBUTING.md)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Review by raduk — APPROVED** [review]
I was wondering if we should add the launcher.py and dependency folder in src

**Comment by rdondera-microsoft** [comment]
> I was wondering if we should add the launcher.py and dependency folder in src

Done.

File: best-practices/largescale-deep-learning/Training/README.md [hunk 1]
```diff
@@ -355,3 +355,7 @@ Nebula Checkpointing improves on standard model checkpointing by saving models 1
   |1|25B|16|	8|	1|	1|	119.42|	4.173	|69.7%|
   |2|20B|8|	8|	1	|1	|117.71	|2.51	|78.5%|
   |3|20B|8|	8|	1	|2|	123.52	|2.63	|80.1%|
+- ### **Pretraining a vision model**
+  Pretraining a vision model can be done in the supervised or the self-supervised learning paradigm. Similar considerations as for language models apply, i.e. pretraining produces good general features, it brings benefits in specialized domains with large amounts of unlabeled data, and it may be costly.
```
Comment (savitamittal1) [inline best-practices/largescale-deep-learning/Training/README.md:359]: can we merge this section under ViT Pretrain?

File: best-practices/largescale-deep-learning/Training/ViT-Pretrain/README.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+3. If network traffic does not exceed the network bandwidth available to the machine, mount the dataset's datastore path and disable caching. This has virtually the same performance as downloading the dataset on the machine when this option is available (disk large enough).
+4. In general, the optimal number of workers for data loading can be found by starting with the number of CPU cores and increasing until there is no noticeable improvement. We have seen that for [NCv3 machines](https://learn.microsoft.com/en-us/azure/virtual-machines/ncv3-series) a good starting point is 6, but for [NCA100v4 machines](https://learn.microsoft.com/en-us/azure/virtual-machines/nc-a100-v4-series) a good starting point is 12.
+5. Scale up before scaling out, e.g. use a Standard_NC24s_v3 with 4 GPUs instead of 4 Standard_NC6s_v3's with 1 GPU each [NCv3 machines](https://learn.microsoft.com/en-us/azure/virtual-machines/ncv3-series). Synchronizing weights on one machine with 4 GPUs (Standard_NC24s_v3) is faster than synchronizing weights across the network between 4 machines with one GPU each (Standard_NC6s_v3).
+6. Use [InfiniBand]#(hardware) whenever possible, as it enables near linear scaling.
```
Comment (savitamittal1) [inline best-practices/largescale-deep-learning/Training/ViT-Pretrain/README.md:17]: Can we provide a link to Azure infiniband VMs available?

File: best-practices/largescale-deep-learning/Training/ViT-Pretrain/README.md [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+### **Multiple nodes / multiple GPUs**
+
+We trained ViTB on ImageNet-1k in a few hardware configurations with varying numbers of nodes and GPUs. The GPU used was V100.
+
+| Nodes | GPUs per Node | Total GPUs | InfiniBand | Training Speed (images/second) |
+|-------|---------------|------------|------------|--------------------------------|
+| 1 | 1 | 1 | no | 101.6 |
+| 1 | 4 | 4 | no | 376.3 |
+| 1 | 4 | 4 | yes | 377.3 |
```
Comment (savitamittal1) [inline best-practices/largescale-deep-learning/Training/ViT-Pretrain/README.md:54]: Should we remove row 3. Its not adding any value here.

File: best-practices/largescale-deep-learning/Training/ViT-Pretrain/README.md (hunk 2 above)
Comment (rdondera-microsoft) [reply best-practices/largescale-deep-learning/Training/ViT-Pretrain/README.md:17]: We have the link in the Hardware section, which point 6 links to. That section also includes details about Infiniband, so keeping things as they are here.

File: best-practices/largescale-deep-learning/Training/ViT-Pretrain/README.md (hunk 3 above)
Comment (rdondera-microsoft) [reply best-practices/largescale-deep-learning/Training/ViT-Pretrain/README.md:54]: Do you mean we already knew that Infiniband on one machine does not make a difference? If we exclude this data point, the Infiniband line in the graph below would start at 16, which would be odd, so I'd rather keep things as they are.

File: best-practices/largescale-deep-learning/Training/README.md (hunk 1 above)
Comment (rdondera-microsoft) [reply best-practices/largescale-deep-learning/Training/README.md:359]: That would make the distinction between language and vision models less visible. How about moving the paragraph under a "Pretraining models" subsection?

File: best-practices/largescale-deep-learning/Training/ViT-Pretrain/README.md (hunk 3 above)
Comment (savitamittal1) [reply best-practices/largescale-deep-learning/Training/ViT-Pretrain/README.md:54]: keeping the row could also confuse people that infiniband is not helping improve the performance.

File: best-practices/largescale-deep-learning/Training/ViT-Pretrain/README.md (hunk 2 above)
Comment (savitamittal1) [reply best-practices/largescale-deep-learning/Training/ViT-Pretrain/README.md:17]: got it. If it provides a link to hardware section, thats good enough.

File: best-practices/largescale-deep-learning/Training/ViT-Pretrain/README.md (hunk 3 above)
Comment (rdondera-microsoft) [reply best-practices/largescale-deep-learning/Training/ViT-Pretrain/README.md:54]: Makes sense. Can clarify this in a note.

**Review by savitamittal1 — APPROVED** [review]

**Review by majercakdavid — APPROVED** [review]
Very nice and informative example!

**Pull request merged**